src/lib/proxy-server.ts

```diff
@@ -35,93 +49,167 @@ async function getSiteByHost( domain: string ): Promise< SiteDetails | null > {
 }
 
 /**
- * Starts the proxy server for sites with custom domains
+ * Common handler for both HTTP and HTTPS requests
  */
-export async function startProxyServer(): Promise< boolean > {
-	if ( isProxyRunning ) return true;
+async function handleProxyRequest(
+	req: http.IncomingMessage,
+	res: http.ServerResponse,
+	isHttps = false
+) {
+	const host = req.headers.host?.split( ':' )[ 0 ]; // Remove port if present
 
-	try {
-		// Create proxy with additional options to preserve host header
-		const proxy = httpProxy.createProxyServer();
-
-		proxy.on( 'error', ( err, req, res ) => {
-			if ( res && res instanceof http.ServerResponse ) {
-				res.writeHead( 500 );
-				res.end( 'Proxy error: ' + err.message );
-			}
+	if ( ! host ) {
+		console.log( 'No host header found' );
+		res.writeHead( 404 );
+		res.end( 'No host header found' );
+		return;
+	}
+
+	const site = await getSiteByHost( host );
+	if ( ! site ) {
+		console.log( `Domain not found: ${ host }` );
+		res.writeHead( 404 );
+		res.end( `Domain not found: ${ host }` );
+		return;
+	}
+
+	if ( ! site.running ) {
+		res.writeHead( 404 );
+		res.end( `The Studio site is currently stopped: ${ site.name }` );
+		return;
+	}
+
+	// If we're on HTTP and site has HTTPS enabled, redirect to HTTPS
+	if ( ! isHttps && site.enableSSL ) {
+		res.writeHead( 301, {
+			Location: `https://${ host }${ req.url }`,
 		} );
+		res.end();
+		return;
+	}
+
+	const headers: Record< string, string > = {};
+
+	if ( isHttps ) {
+		headers[ 'X-Forwarded-Proto' ] = 'https';
+	}
 
-		proxyServer = http.createServer( async ( req, res ) => {
-			const host = req.headers.host?.split( ':' )[ 0 ]; // Remove port if present
-
-			// Look up the port directly from user data
-			if ( ! host ) {
-				console.log( 'No host header found' );
-				res.writeHead( 404 );
-				res.end( 'No host header found' );
-				return;
-			}
-
-			const site = await getSiteByHost( host );
-			if ( ! site ) {
-				console.log( `Domain not found: ${ host }` );
-				res.writeHead( 404 );
-				res.end( `Domain not found: ${ host }` );
-				return;
-			}
-
-			if ( ! site.running ) {
-				res.writeHead( 404 );
-				res.end( `The Studio site is currently stopped: ${ site.name }` );
-				return;
-			}
-
-			// Forward the request with the original host preserved
-			proxy.web( req, res, {
-				target: `http://localhost:${ site.port }`,
-				xfwd: true, // Pass along x-forwarded headers
+	proxy.web( req, res, {
+		target: `http://localhost:${ site.port }`,
+		xfwd: true, // Pass along x-forwarded headers
+		headers,
+	} );
+}
+
+export async function checkPortInWindows( port: number ): Promise< boolean > {
+	// On Windows, node doesn't throw an error if port is busy, so we use the net module to explicitly check
+	// if it's possible to establish a TCP connection to that port (meaning it's busy)
+	if ( process.platform === 'win32' ) {
```

Comment:
```suggestion
// On Windows, node doesn't throw an error if port is busy, so we use the net module to explicitly check
// if it's possible to establish a TCP connection to that port (meaning it's busy)
export async function checkPortOnWindows( port: number ): Promise< boolean > {
	if ( process.platform !== 'win32' ) {
		return true;
	}
```
Nit, but the comment placement and the function name could be tweaked. Seems logical to return early instead of having a big `if` wrapper, too.